### PR TITLE
[tf.data] Support eager mode benchmarking

### PR DESCRIPTION
**Pull request by kvignesh1420**
This PR adds support for eager execution based benchmarks based on the benchmarking context. This is achieved by checking for the `eager_execution()` value within the benchmarking context and iterating over the dataset eagerly in the `DatasetBenchmarkBase` class.

Additionally, the `ListFilesBenchmark` in `python/data/benchmarks/list_files_benchmark.py` has been refactored to use the `DatasetBenchmarkBase` class.

A sample comparison w.r.t `RangeBenchmark` is as follows:

GRAPH MODE:
```python
entry {
  name: "RangeBenchmark.modeling_on.graph"
  iters: 5
  wall_time: 1.7887322902679443e-07
  extras {
    key: "num_elements"
    value {
      double_value: 10000000.0
    }
  }
}

entry {
  name: "RangeBenchmark.modeling_off.graph"
  iters: 5
  wall_time: 1.541191530227661e-07
  extras {
    key: "num_elements"
    value {
      double_value: 50000000.0
    }
  }
}
```

EAGER MODE:
```python
entry {
  name: "RangeBenchmark.modeling_on.eager"
  iters: 5
  wall_time: 2.0390429496765137e-07
  extras {
    key: "num_elements"
    value {
      double_value: 10000000.0
    }
  }
}

entry {
  name: "RangeBenchmark.modeling_off.eager"
  iters: 5
  wall_time: 1.8014323711395263e-07
  extras {
    key: "num_elements"
    value {
      double_value: 50000000.0
    }
  }
}
```

w.r.t  `ListFilesBenchmark`
```python

GRAPH MODE
entry {
  name: "ListFilesBenchmark.nested_directory(1024*16).graph"
  iters: 3
  wall_time: 8.092029020190239e-07
  extras {
    key: "num_elements"
    value {
      double_value: 2048.0
    }
  }
}

EAGER MODE
entry {
  name: "ListFilesBenchmark.nested_directory(1024*16).eager"
  iters: 3
  wall_time: 7.745111361145973e-07
  extras {
    key: "num_elements"
    value {
      double_value: 2048.0
    }
  }
}
```

NOTE: Description has been updated as per review comments.

### CODE REVIEW

**Comment by kvignesh1420** [comment]
cc: @aaudiber

**Comment by aaudiber** [comment]
I don't think eager benchmarks are worth maintaining, since the backend dataset implementation works the same in both eager and graph mode. WDYT @jsimsa?

**Comment by jsimsa** [comment]
I think there is value in being able to run benchmarks in eager mode but this support should certainly not result in duplicating benchmark code. Instead, the benchmark base should be modified to make it possible to run a benchmark both in graph mode or eagerly. Which mode to use could be a based on which parameter is passed to `python/data/benchmarks/benchmark_base.py` `run_benchmark` method, defaulting to the mode of the context in which the benchmark is executed.

**Comment by kvignesh1420** [comment]
I see that not all the benchmarks located in `python/data/benchmarks/` and `python/data/experimental/benchmarks/` subclass the `DatasetBenchmarkBase` class. Thus, making changes to the `DatasetBenchmarkBase` class will cover a few of them but not all. This is due to the custom `_benchmark` logic that is needed instead of the default one available in `DatasetBenchmarkBase`.

As per @jsimsa 's suggestion, I can modify the `run_benchmark` method of `DatasetBenchmarkBase` class to take a parameter named `eager` which will either be `True` or `False`. It will default to `False` so that it doesn't disturb the current benchmark scenarios. I can then add conditional logic to handle `eager=True` case within that method. The benchmarks which inherit the `DatasetBenchmarkBase` can now call the `run_and_report_benchmark` method with the option of benchmarking eagerly.

I can do the same for benchmarks which do not inherit the `DatasetBenchmarkBase` class and modify the `_benchmark` method to take the `eager` parameter and conditionalize the logic.

What do you think? @jsimsa @aaudiber?

**Comment by jsimsa** [comment]
Thanks @kvignesh1420. You plan for modifying `DatasetBenchmarkBase` sounds good to me.  It would also be good to understand why do some benchmarks need custom `_benchmark` logic and whether it would be possible to refactor things so that all benchmarks inherit from `DatasetBenchmarkBase`.

**Comment by kvignesh1420** [comment]
@jsimsa thanks for the go-ahead. I will update the PR with the required changes and try to refactor things.

**Comment by kvignesh1420** [comment]
I have modified the `DatasetBenchmarkBase` class to support eager iteration over the datasets depending on the context. Additionally, the ListFilesBenchmark in python/data/benchmarks/list_files_benchmark.py has been refactored to use the `DatasetBenchmarkBase` class. The refactor changes the current output of the benchmark in order to be in sync with the style of other benchmarks.

cc: @jsimsa @aaudiber let me know how it looks. I will take up the benchmarks in `data/experimental/benchmarks` after this.

File: tensorflow/python/data/benchmarks/benchmark_base.py [hunk 1]
```diff
@@ -60,6 +61,20 @@ def run_benchmark(self,
     options.experimental_optimization.apply_default_optimizations = (
         apply_default_optimizations)
     dataset = dataset.with_options(options)
+
```
Comment (jsimsa) [inline tensorflow/python/data/benchmarks/benchmark_base.py:64]: lines 78-82 should be moved here

**Review by jsimsa — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/benchmarks/benchmark_base.py (hunk 1 above)
Comment (kvignesh1420) [reply tensorflow/python/data/benchmarks/benchmark_base.py:66]: addressed in the commit.

File: tensorflow/python/data/benchmarks/benchmark_base.py [hunk 2]
```diff
@@ -56,15 +56,32 @@ def run_benchmark(self,
...
+          next(iterator)
+        start = time.time()
+        next(iterator)
+        end = time.time()
+        deltas.append(end - start)
+      return np.median(deltas) / float(num_elements)
+
     iterator = dataset_ops.make_initializable_iterator(dataset)
     next_element = iterator.get_next()
     next_element = nest.flatten(next_element)[0]
```
Comment (kvignesh1420) [inline tensorflow/python/data/benchmarks/benchmark_base.py:87]: @jsimsa is this transformation necessary? I have not added this in the eager case.

File: tensorflow/python/data/benchmarks/benchmark_base.py [hunk 3]
```diff
@@ -56,15 +56,34 @@ def run_benchmark(self,
...
+        iterator = iter(dataset)
+        start = time.time()
+        next(iterator)
+        end = time.time()
+        deltas.append(end - start)
+      return np.median(deltas) / float(num_elements)
+
     iterator = dataset_ops.make_initializable_iterator(dataset)
     next_element = iterator.get_next()
     next_element = nest.flatten(next_element)[0]
```
Comment (kvignesh1420) [inline tensorflow/python/data/benchmarks/benchmark_base.py:89]: @jsimsa  Is the nest.flatten() transformation necessary? I have not added this in the eager case.

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thank you. As per one of my comments, we want to test that the new implementation works as expected before merging the PR (and I provided you with instructions on how to do so).

File: tensorflow/python/data/benchmarks/benchmark_base.py [hunk 4]
```diff
@@ -56,15 +56,34 @@ def run_benchmark(self,
...
+    options = dataset.options()
     options.experimental_optimization.apply_default_optimizations = (
         apply_default_optimizations)
     dataset = dataset.with_options(options)
+
     # NOTE: We use `dataset.skip()` to perform the iterations in C++, avoiding
-    # the overhead of multiple `session.run()` calls. Note that this relies on
-    # the underlying implementation of `skip`: if it is optimized in the future,
-    # we will have to change this code.
+    # the overhead of multiple `session.run()` calls (in graph mode). Note that
```
[the remark this answers is not in the document]
Comment (kvignesh1420) [reply tensorflow/python/data/benchmarks/benchmark_base.py]: addressed in the commit

File: tensorflow/python/data/benchmarks/benchmark_base.py (hunk 3 above)
Comment (kvignesh1420) [reply tensorflow/python/data/benchmarks/benchmark_base.py:90]: @jsimsa setting/unsetting the `TF2_BEHAVIOR` env var didn't make a difference. I had to change the .bazelrc file and modify the config from:
```
build --config=v2
test --config=v2
```
to:
```
build --config=v1
test --config=v1
```
Then it picked up the config saying:
```
Found applicable config definition build:v1 in file /home/vikoth18/tensorflow/.bazelrc: --define=tf_api_version=1 --action_env=TF2_BEHAVIOR=0
```
If there is an alternative or a better approach please let me know. I am currently running the benchmarks and will post the outputs soon.

**Comment by kvignesh1420** [comment]
@jsimsa I have modified the `run_and_report_benchmark` method in `DatasetBenchmarkBase` to append the mode in which the benchmarks are running. Also, a sample benchmark result has been added to the description. If you want additional results from eager mode benchmarks and its graph counter-part, I can attach the files with the log outputs. Let me know.

File: tensorflow/python/data/benchmarks/benchmark_base.py [hunk 5]
```diff
@@ -95,7 +115,10 @@ def run_and_report_benchmark(self,
     # Measure the per-element wall time.
     wall_time = self.run_benchmark(dataset, num_elements, iters, warmup,
                                    apply_default_optimizations)
-
+    if context.executing_eagerly():
+      name = "{}.EagerMode".format(name)
```
Comment (jsimsa) [inline tensorflow/python/data/benchmarks/benchmark_base.py:119]: replace "EagerMode" with "eager" and "GraphMode" and "graph"

**Review by jsimsa — COMMENTED** [review]
This looks great. Thank you for the changes. Left one minor comment. Other than that it looks good.

File: tensorflow/python/data/benchmarks/benchmark_base.py (hunk 5 above)
Comment (kvignesh1420) [reply tensorflow/python/data/benchmarks/benchmark_base.py]: @jsimsa addressed it in the commit.

**Review by jsimsa — APPROVED** [review]

**Pull request merged**